### ISSUE TITLE
Tablet migration and view update from staging don't interact well

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylladb#19149]
There are two problems

1) Mutation-based streaming is mixing staging and non-staging data. The receiver relies on distributed table view_status to determine if a sstable should be registered into generator (see db::view::check_needs_view_update_path()). If the generator hasn't started processing the base in the leaving replica before migration happens (generator processes one table at a time, so that's possible in practice), no sstables streamed to pending replica will be registered, causing base-view inconsistency. Otherwise, all the sstables will be registered into the generator, causing a significant amount of extra view update work.

2) SSTables are being registered into generator prematurely, on pending replica. There's a window between migration's streaming and read_from_new stages, and during that window, view updates will be ignored in the new pending replica (see db::view::get_view_natural_endpoint()), again causing base-view inconsistency.

**Comment by bhalevy** [comment]
@raphaelsc please separate the problem definition, which should be in the issue description, from the proposed solution, that should be described in a separate comment, so we can possibly discuss alternatives, that may solve the isuue slightly differently.

**Comment by raphaelsc** [comment]
Possible solution for # 1:

Can be fixed by doing the streaming in two stages. First one will stream staging sstables (using stream_reason::repair), so they'll be registered into the generator. The second will stream non-staging data as today, using stream_reason::tablet_migration, which will bypass their registration. We can possibly special case stream_reason::tablet_migration within db::view::check_needs_view_update_path() knowing that all sstables produced with that reason will be non-staging data. stream_reason::repair is special cased today but for the opposite reason, all sstables produced with that reason are considered staging.

Possible solution for # 2:

Can be fixed by postponing registration in new replica until it's no longer considered pending. The registration can be done by scanning the sstable set and registering those in staging directory.

**Comment by raphaelsc** [comment]
Tip when testing: error view_update_generator_consume_staging_sstable can be injected to block view updates from staging on leaving replica, allowing the test to verify the backlog will be correctly transferred on migration.

**Comment by bhalevy** [comment]
After discussion tombstone_gc in views with @tgrabiec we came to a few related observations.
Cc @denesb @asias @raphaelsc @nyh @piodul @karol-kokoszka 
* Since repair of a base table may cause view update generation when streamed sstables in staging are processed, we should declare repair as done (with the effective time_point of when it started) only when the respective view update generation phase is complete.
  * Can this be achieved by doing/syncing with the view update generation as a state in the tablet repair transition state machine?
* Views should be repaired after the base table is repaired (and view update generation is done, as per the above)
  * Need to verify if scylla-manger is already doing it today.
  * Does it happen in scylla when repair the whole keyspace?

**Comment by karol-kokoszka** [comment]
> Views should be repaired after the base table is repaired (and view
update generation is done, as per the above)
>        Need to verify if scylla-manger is already doing it today.

Yes, scylla manager is repairing views after the base table.
However, the CQL credentials must be provided to the manager, as we are
querying `system_schema.views` to list available views.

https://github.com/scylladb/scylla-manager/issues/3367

**Comment by karol-kokoszka** [comment]
Here is the doc describing current state of the repair in Scylla Manager https://docs.google.com/document/d/1mBaufbomXU6hDO_25KkhC7CO65AbFgKpHVpB5Mei8xA/edit?usp=sharing

**Comment by bhalevy** [comment]
Note that even if we keep the staging state of migrated sstables, those need to be registered with the view update generator on the target replica only when migration is done, since view update generation needs to compute the delta between the rows state with vs. without staging in order to generate the view update.  This can be done only after all data has been migrated.

**Comment by raphaelsc** [comment]
> Note that even if we keep the staging state of migrated sstables, those need to be registered with the view update generator on the target replica only when migration is done, since view update generation needs to compute the delta between the rows state with vs. without staging in order to generate the view update. This can be done only after all data has been migrated.

Exactly. That's problem #2 in the description. Today, view update path will ignore those update attempts on pending, so update generated on partial state will not be pushed to corresponding view replica. But this is problematic, because it cause us to miss view updates for some tokens, since the sstable might be moved out of staging once it's fully processed in the generator. The ignoring of update on pending is silent, so things proceed as if nothing wrong happened.

**Comment by mykaul** [comment]
@bhalevy - is this a 6.2 material?

**Comment by bhalevy** [comment]
> @bhalevy - is this a 6.2 material?

Unfortunately not.  It looks like we're going to miss that train.
